c++20/libstdc++: `char_traits.h:213:13: note: subexpression not valid in a constant expression`

using this constexpr boost static_string (requires static_string develop) does not compile with clang against libstdc++ (g++ compiles against libstdc++ also clang++ compiles against libc++).

```c++
#include <boost/static_string.hpp>
using str = boost::static_strings::static_string<38>;
constexpr str a{ "mystr" };
```

```
c.cpp:5:15: error: constexpr variable 'a' must be initialized by a constant expression
constexpr str a{ "mystr" };
              ^~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/char_traits.h:213:13: note: subexpression not valid in a constant expression
   if (__s1 == __s2)
            ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/char_traits.h:417:11: note: in call to 'move(&a.static_string_base::data_[0], &"mystr"[0], 5)'
   return __gnu_cxx::char_traits<char_type>::move(__s1, __s2, __n);
          ^
/home/tim/NIBuild/3rdparty/boost-v1.78.0-R8/boost/static_string/static_string.hpp:6000:3: note: in call to 'move(&a.static_string_base::data_[0], &"mystr"[0], 5)'
  traits_type::move(data(), s, size());
  ^
/home/tim/NIBuild/3rdparty/boost-v1.78.0-R8/boost/static_string/static_string.hpp:1413:12: note: in call to '&a->assign(&"mystr"[0], 5)'
    return assign(s, traits_type::length(s));
           ^
/home/tim/NIBuild/3rdparty/boost-v1.78.0-R8/boost/static_string/static_string.hpp:961:5: note: in call to '&a->assign(&"mystr"[0])'
    assign(s);
    ^
c.cpp:5:15: note: in call to 'basic_static_string(&"mystr"[0])'
constexpr str a{ "mystr" };
              ^
1 error generated.
```

the offending code in libstdc++ is this one:
```
	  if (__s1 == __s2) // unlikely, but saves a lot of work
	    return __s1;
```
https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/include/bits/char_traits.h#L216-L217

--- a/clang/lib/AST/ExprConstant.cpp
+++ b/clang/lib/AST/ExprConstant.cpp
@@ -2142,7 +2142,7 @@ static bool IsLiteralLValue(const LValue &Value) {
   if (Value.getLValueCallIndex())
     return false;
   const Expr *E = Value.Base.dyn_cast<const Expr*>();
-  return E && !isa<MaterializeTemporaryExpr>(E);
+  return E && !isa<MaterializeTemporaryExpr, StringLiteral>(E);
 }
 
 static bool IsWeakLValue(const LValue &Value) {

--- a/clang/test/AST/ByteCode/cxx20.cpp
+++ b/clang/test/AST/ByteCode/cxx20.cpp
@@ -108,22 +108,16 @@ constexpr auto p2 = "test2";
 constexpr bool b1 = foo(p1) == foo(p1);
 static_assert(b1);
 
-constexpr bool b2 = foo(p1) == foo(p2); // ref-error {{must be initialized by a constant expression}} \
-                                        // ref-note {{comparison of addresses of literals}} \
-                                        // ref-note {{declared here}}
-static_assert(!b2); // ref-error {{not an integral constant expression}} \
-                    // ref-note {{not a constant expression}}
+constexpr bool b2 = foo(p1) == foo(p2);
+static_assert(!b2);
 
 constexpr auto name1() { return "name1"; }
 constexpr auto name2() { return "name2"; }
 
 constexpr auto b3 = name1() == name1();
 static_assert(b3);
-constexpr auto b4 = name1() == name2(); // ref-error {{must be initialized by a constant expression}} \
-                                        // ref-note {{has unspecified value}} \
-                                        // ref-note {{declared here}}
-static_assert(!b4); // ref-error {{not an integral constant expression}} \
-                    // ref-note {{not a constant expression}}
+constexpr auto b4 = name1() == name2();
+static_assert(!b4);
 
 namespace UninitializedFields {
   class A {

--- a/clang/test/SemaCXX/constant-expression-cxx11.cpp
+++ b/clang/test/SemaCXX/constant-expression-cxx11.cpp
@@ -358,11 +358,9 @@ struct Str {
 
 extern char externalvar[];
 constexpr bool constaddress = (void *)externalvar == (void *)0x4000UL; // expected-error {{must be initialized by a constant expression}} expected-note {{reinterpret_cast}}
-constexpr bool litaddress = "foo" == "foo"; // expected-error {{must be initialized by a constant expression}}
-// expected-note@-1 {{comparison of addresses of literals has unspecified value}}
-// cxx20_23-warning@-2 {{comparison between two arrays is deprecated}}
+constexpr bool litaddress = "foo" == "foo"; // cxx20_23-warning {{comparison between two arrays is deprecated}}
 static_assert(0 != "foo", "");
-
+static_assert("foo" != "foo", "");// cxx20_23-warning {{comparison between two arrays is deprecated}}
 }
 
 namespace MaterializeTemporary {

